File: tools/build_ftc_validation_workbook.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const rootDir = process.cwd();
const outputsDir = path.join(rootDir, "outputs");
const validationDir = path.join(outputsDir, "ftc_validation");
const previewDir = path.join(validationDir, "previews");
const candidateCsvPath = path.join(outputsDir, "ftc_candidate_table.csv");
const selectedCsvPath = path.join(outputsDir, "ftc_selected_clusters.csv");
const commentsCsvPath = path.join(outputsDir, "comment_topics_clusters.csv");
const outputPath = path.join(validationDir, "ftc_validasi_entropy_overlap.xlsx");

function parseCsv(text) {
  const rows = [];
  let row = [];
  let cell = "";
  let inQuotes = false;

  for (let i = 0; i < text.length; i += 1) {
    const char = text[i];
    const next = text[i + 1];

    if (char === '"' && inQuotes && next === '"') {
      cell += '"';
      i += 1;
    } else if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === "," && !inQuotes) {
      row.push(cell);
      cell = "";
    } else if ((char === "\n" || char === "\r") && !inQuotes) {
      if (char === "\r" && next === "\n") {
        i += 1;
      }
      row.push(cell);
      if (row.some((value) => value !== "")) {
        rows.push(row);
      }
      row = [];
      cell = "";
    } else {
      cell += char;
    }
  }

  if (cell.length > 0 || row.length > 0) {
    row.push(cell);
    rows.push(row);
  }

  const [headers, ...dataRows] = rows;
  return dataRows.map((values) => Object.fromEntries(headers.map((header, index) => [header, values[index] ?? ""])));
}

function colName(columnNumberOneBased) {
  let value = columnNumberOneBased;
  let name = "";
  while (value > 0) {
    const remainder = (value - 1) % 26;
    name = String.fromCharCode(65 + remainder) + name;
    value = Math.floor((value - 1) / 26);
  }
  return name;
}

function parseDocIds(clusterCandidate) {
  return [...clusterCandidate.matchAll(/D(\d+)/g)].map((match) => Number(match[1]));
}

function toNumber(value) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : 0;
}

function writeMatrix(sheet, startRow, startCol, matrix) {
  if (!matrix.length || !matrix[0].length) return;
  sheet.getRangeByIndexes(startRow, startCol, matrix.length, matrix[0].length).values = matrix;
}

function writeFormulas(sheet, startRow, startCol, matrix) {
  if (!matrix.length || !matrix[0].length) return;
  sheet.getRangeByIndexes(startRow, startCol, matrix.length, matrix[0].length).formulas = matrix;
}

function styleHeader(range) {
  range.format = {
    fill: "#173B7A",
    font: { bold: true, color: "#FFFFFF" },
    wrapText: true,
    horizontalAlignment: "center",
    verticalAlignment: "middle",
  };
}

function styleTable(range) {
  range.format = {
    borders: { preset: "all", style: "thin", color: "#C9D4E5" },
    verticalAlignment: "top",
    wrapText: true,
  };
}

await fs.mkdir(previewDir, { recursive: true });

const candidateRows = parseCsv(await fs.readFile(candidateCsvPath, "utf8"));
const selectedRows = parseCsv(await fs.readFile(selectedCsvPath, "utf8"));
const commentRows = parseCsv(await fs.readFile(commentsCsvPath, "utf8"));

const iterationOne = candidateRows
  .filter((row) => Number(row.iteration) === 1)
  .sort((a, b) => Number(a.no_cluster) - Number(b.no_cluster));

const docCount = commentRows.length;
const candidateCount = iterationOne.length;
const docStartColIndex = 5; // F
const docStartCol = colName(docStartColIndex + 1);
const docEndCol = colName(docStartColIndex + docCount);
const candidateStartRow = 8;
const candidateEndRow = candidateStartRow + candidateCount - 1;

const workbook = Workbook.create();
const candidatesSheet = workbook.worksheets.add("FTC Candidates");
const validationSheet = workbook.worksheets.add("EO Validation");
const selectedSheet = workbook.worksheets.add("Selected Clusters");

for (const sheet of [candidatesSheet, validationSheet, selectedSheet]) {
  sheet.showGridLines = false;
}

// Sheet 1: table like the FTC slide.
candidatesSheet.getRange("A1:G1").merge();
candidatesSheet.getRange("A1").values = [["Tahap 1. Kemunculan frequent term pada beberapa dokumen"]];
candidatesSheet.getRange("A1").format = {
  fill: "#173B7A",
  font: { bold: true, color: "#FFFFFF", size: 16 },
};
candidatesSheet.getRange("A2:G2").merge();
candidatesSheet.getRange("A2").values = [[`FTC Iterasi 1, minimum support = 8 dokumen, jumlah kandidat = ${candidateCount}`]];

writeMatrix(candidatesSheet, 3, 0, [[
  "No Cluster",
  "Frequent term set",
  "Cluster candidate",
  "Support",
  "EO Python",
  "EO Excel",
  "Selisih",
]]);

const candidateTableValues = iterationOne.map((row) => [
  Number(row.no_cluster),
  row.frequent_term_set,
  row.cluster_candidate,
  toNumber(row.support),
  toNumber(row.eo),
  null,
  null,
]);
writeMatrix(candidatesSheet, 4, 0, candidateTableValues);

const candidateFormulaRows = iterationOne.map((_, index) => {
  const excelRow = 5 + index;
  const validationRow = candidateStartRow + index;
  return [
    `='EO Validation'!E${validationRow}`,
    `=ABS(E${excelRow}-F${excelRow})`,
  ];
});
writeFormulas(candidatesSheet, 4, 5, candidateFormulaRows);

candidatesSheet.getRange("A4:G4").format = {
  fill: "#173B7A",
  font: { bold: true, color: "#FFFFFF" },
  horizontalAlignment: "center",
  wrapText: true,
};
styleTable(candidatesSheet.getRange(`A4:G${4 + candidateCount}`));
candidatesSheet.getRange("A:A").format.columnWidthPx = 82;
candidatesSheet.getRange("B:B").format.columnWidthPx = 180;
candidatesSheet.getRange("C:C").format.columnWidthPx = 620;
candidatesSheet.getRange("D:G").format.columnWidthPx = 90;
candidatesSheet.getRange(`D5:G${4 + candidateCount}`).format.numberFormat = "0.000000";
candidatesSheet.freezePanes.freezeRows(4);

// Sheet 2: formula validation for EO.
validationSheet.getRange("A1:E1").merge();
validationSheet.getRange("A1").values = [["Tahap 2. Penerapan Persamaan Entropy Overlap (EO)"]];
validationSheet.getRange("A1").format = {
  fill: "#173B7A",
  font: { bold: true, color: "#FFFFFF", size: 16 },
};
validationSheet.getRange("A2:E2").merge();
validationSheet.getRange("A2").values = [[
  "EO dihitung dengan kontribusi per dokumen: ((-1/f_d) * LN(1/f_d)); f_d adalah jumlah kandidat cluster yang memuat dokumen D tersebut.",
]];

writeMatrix(validationSheet, 3, 0, [["Frekuensi dokumen f_d", "", "", "", ""]]);
writeMatrix(validationSheet, 4, 0, [["Kontribusi EO per dokumen", "", "", "", ""]]);

const docHeaders = Array.from({ length: docCount }, (_, index) => `D${index + 1}`);
writeMatrix(validationSheet, 3, docStartColIndex, [docHeaders]);

const frequencyFormulas = docHeaders.map((_, index) => {
  const column = colName(docStartColIndex + index + 1);
  return `=SUM(${column}$${candidateStartRow}:${column}$${candidateEndRow})`;
});
const contributionFormulas = docHeaders.map((_, index) => {
  const column = colName(docStartColIndex + index + 1);
  return `=IF(${column}$4>1,((-1/${column}$4)*LN(1/${column}$4)),0)`;
});
writeFormulas(validationSheet, 3, docStartColIndex, [frequencyFormulas]);
writeFormulas(validationSheet, 4, docStartColIndex, [contributionFormulas]);

writeMatrix(validationSheet, 6, 0, [[
  "No Cluster",
  "Frequent term set",
  "Cluster candidate",
  "Support",
  "EO Excel",
  ...docHeaders,
]]);

const membershipRows = iterationOne.map((row) => {
  const docs = new Set(parseDocIds(row.cluster_candidate));
  return [
    Number(row.no_cluster),
    row.frequent_term_set,
    row.cluster_candidate,
    null,
    null,
    ...Array.from({ length: docCount }, (_, index) => (docs.has(index + 1) ? 1 : 0)),
  ];
});
writeMatrix(validationSheet, 7, 0, membershipRows);

const validationFormulas = iterationOne.map((_, index) => {
  const excelRow = candidateStartRow + index;
  return [
    `=SUM(${docStartCol}${excelRow}:${docEndCol}${excelRow})`,
    `=SUMPRODUCT(${docStartCol}${excelRow}:${docEndCol}${excelRow},$${docStartCol}$5:$${docEndCol}$5)`,
  ];
});
writeFormulas(validationSheet, 7, 3, validationFormulas);

styleHeader(validationSheet.getRange(`A7:${docEndCol}7`));
styleTable(validationSheet.getRange(`A7:${docEndCol}${candidateEndRow}`));
validationSheet.getRange("A:A").format.columnWidthPx = 82;
validationSheet.getRange("B:B").format.columnWidthPx = 180;
validationSheet.getRange("C:C").format.columnWidthPx = 620;
validationSheet.getRange("D:E").format.columnWidthPx = 95;
validationSheet.getRangeByIndexes(0, docStartColIndex, candidateEndRow, docCount).format.columnWidthPx = 38;
validationSheet.getRange(`D8:E${candidateEndRow}`).format.numberFormat = "0.000000";
validationSheet.getRange(`${docStartCol}4:${docEndCol}5`).format.numberFormat = "0.000000";
validationSheet.freezePanes.freezeRows(7);
validationSheet.freezePanes.freezeColumns(5);

// Sheet 3: selected clusters per FTC iteration.
selectedSheet.getRange("A1:F1").merge();
selectedSheet.getRange("A1").values = [["Cluster Terpilih FTC per Iterasi"]];
selectedSheet.getRange("A1").format = {
  fill: "#173B7A",
  font: { bold: true, color: "#FFFFFF", size: 16 },
};
writeMatrix(selectedSheet, 3, 0, [[
  "Iteration",
  "Cluster ID",
  "Frequent termset",
  "Support",
  "Entropy Overlap",
  "Remaining docs after",
]]);
writeMatrix(selectedSheet, 4, 0, selectedRows.map((row) => [
  Number(row.iteration),
  Number(row.cluster_id),
  row.frequent_termset,
  toNumber(row.support),
  toNumber(row.entropy_overlap),
  toNumber(row.n_remaining_docs_after),
]));
styleHeader(selectedSheet.getRange("A4:F4"));
styleTable(selectedSheet.getRange(`A4:F${4 + selectedRows.length}`));
selectedSheet.getRange("A:B").format.columnWidthPx = 90;
selectedSheet.getRange("C:C").format.columnWidthPx = 220;
selectedSheet.getRange("D:F").format.columnWidthPx = 120;
selectedSheet.getRange(`E5:E${4 + selectedRows.length}`).format.numberFormat = "0.000000";
selectedSheet.freezePanes.freezeRows(4);

for (const [sheetName, range, fileName] of [
  ["FTC Candidates", "A1:G25", "ftc_candidates_preview.png"],
  ["EO Validation", "A1:K25", "eo_validation_preview.png"],
  ["Selected Clusters", "A1:F25", "selected_clusters_preview.png"],
]) {
  const preview = await workbook.render({ sheetName, range, scale: 1, format: "png" });
  await fs.writeFile(path.join(previewDir, fileName), new Uint8Array(await preview.arrayBuffer()));
}

const inspection = await workbook.inspect({
  kind: "table",
  sheetId: "FTC Candidates",
  range: "A1:G12",
  include: "values,formulas",
  tableMaxRows: 12,
  tableMaxCols: 7,
});
console.log(inspection.ndjson);

const formulaErrors = await workbook.inspect({
  kind: "match",
  searchTerm: "#REF!|#DIV/0!|#VALUE!|#NAME\\?|#N/A",
  options: { useRegex: true, maxResults: 200 },
  summary: "final formula error scan",
});
console.log(formulaErrors.ndjson);

const xlsx = await SpreadsheetFile.exportXlsx(workbook);
await xlsx.save(outputPath);
console.log(outputPath);

File: tools/build_ftc_validation_workbook.py
from __future__ import annotations

import csv
import math
import re
from pathlib import Path

import openpyxl
import xlsxwriter


ROOT_DIR = Path(__file__).resolve().parents[1]
OUTPUTS_DIR = ROOT_DIR / "outputs"
VALIDATION_DIR = OUTPUTS_DIR / "ftc_validation"
WORKBOOK_PATH = VALIDATION_DIR / "ftc_validasi_entropy_overlap.xlsx"

CANDIDATE_CSV = OUTPUTS_DIR / "ftc_candidate_table.csv"
SELECTED_CSV = OUTPUTS_DIR / "ftc_selected_clusters.csv"
COMMENTS_CSV = OUTPUTS_DIR / "comment_topics_clusters.csv"


def read_csv(path: Path) -> list[dict[str, str]]:
    with path.open("r", encoding="utf-8", newline="") as file:
        return list(csv.DictReader(file))


def parse_doc_ids(cluster_candidate: str) -> list[int]:
    return [int(match.group(1)) for match in re.finditer(r"D(\d+)", cluster_candidate)]


def excel_col_name(col_idx_zero_based: int) -> str:
    value = col_idx_zero_based + 1
    name = ""
    while value:
        value, remainder = divmod(value - 1, 26)
        name = chr(65 + remainder) + name
    return name


def as_int(value: str) -> int:
    return int(float(value or 0))


def as_float(value: str) -> float:
    return float(value or 0)


def build_workbook() -> Path:
    VALIDATION_DIR.mkdir(parents=True, exist_ok=True)

    candidate_rows = read_csv(CANDIDATE_CSV)
    selected_rows = read_csv(SELECTED_CSV)
    comment_rows = read_csv(COMMENTS_CSV)

    iteration_one = sorted(
        [row for row in candidate_rows if as_int(row["iteration"]) == 1],
        key=lambda row: as_int(row["no_cluster"]),
    )

    doc_count = len(comment_rows)
    candidate_count = len(iteration_one)
    doc_start_col = 5  # F
    doc_end_col = doc_start_col + doc_count - 1
    doc_start_name = excel_col_name(doc_start_col)
    doc_end_name = excel_col_name(doc_end_col)
    candidate_start_excel_row = 8
    candidate_end_excel_row = candidate_start_excel_row + candidate_count - 1

    memberships: list[set[int]] = [
        set(parse_doc_ids(row["cluster_candidate"])) for row in iteration_one
    ]
    doc_frequencies = {
        doc_id: sum(1 for member_docs in memberships if doc_id in member_docs)
        for doc_id in range(1, doc_count + 1)
    }
    doc_contributions = {
        doc_id: (math.log(freq) / freq if freq > 1 else 0.0)
        for doc_id, freq in doc_frequencies.items()
    }
    excel_eo_values = [
        sum(doc_contributions[doc_id] for doc_id in member_docs)
        for member_docs in memberships
    ]

    workbook = xlsxwriter.Workbook(WORKBOOK_PATH)
    workbook.set_calc_mode("auto")

    title_fmt = workbook.add_format(
        {"bold": True, "font_color": "white", "bg_color": "#173B7A", "font_size": 16}
    )
    subtitle_fmt = workbook.add_format({"italic": True, "font_color": "#26364D"})
    header_fmt = workbook.add_format(
        {
            "bold": True,
            "font_color": "white",
            "bg_color": "#173B7A",
            "align": "center",
            "valign": "vcenter",
            "border": 1,
            "text_wrap": True,
        }
    )
    text_fmt = workbook.add_format({"border": 1, "valign": "top", "text_wrap": True})
    number_fmt = workbook.add_format({"border": 1, "num_format": "0.000000"})
    integer_fmt = workbook.add_format({"border": 1, "num_format": "0"})
    formula_fmt = workbook.add_format(
        {"border": 1, "num_format": "0.000000", "bg_color": "#F8FBFF"}
    )
    member_fmt = workbook.add_format({"border": 1, "num_format": "0", "align": "center"})

    # Sheet 1: FTC candidates, close to the lecture-table format.
    ws_candidates = workbook.add_worksheet("FTC Candidates")
    ws_candidates.hide_gridlines(2)
    ws_candidates.merge_range(
        "A1:G1", "Tahap 1. Kemunculan frequent term pada beberapa dokumen", title_fmt
    )
    ws_candidates.merge_range(
        "A2:G2",
        f"FTC Iterasi 1, minimum support = 8 dokumen, jumlah kandidat = {candidate_count}",
        subtitle_fmt,
    )
    ws_candidates.write_row(
        "A4",
        ["No Cluster", "Frequent term set", "Cluster candidate", "Support", "EO Python", "EO Excel", "Selisih"],
        header_fmt,
    )

    for idx, row in enumerate(iteration_one, start=0):
        excel_row = 5 + idx
        validation_row = candidate_start_excel_row + idx
        eo_python = as_float(row["eo"])
        eo_excel_value = excel_eo_values[idx]
        ws_candidates.write_number(excel_row - 1, 0, as_int(row["no_cluster"]), integer_fmt)
        ws_candidates.write(excel_row - 1, 1, row["frequent_term_set"], text_fmt)
        ws_candidates.write(excel_row - 1, 2, row["cluster_candidate"], text_fmt)
        ws_candidates.write_number(excel_row - 1, 3, as_int(row["support"]), integer_fmt)
        ws_candidates.write_number(excel_row - 1, 4, eo_python, number_fmt)
        ws_candidates.write_formula(
            excel_row - 1,
            5,
            f"='EO Validation'!E{validation_row}",
            formula_fmt,
            eo_excel_value,
        )
        ws_candidates.write_formula(
            excel_row - 1,
            6,
            f"=ABS(E{excel_row}-F{excel_row})",
            formula_fmt,
            abs(eo_python - eo_excel_value),
        )

    ws_candidates.set_column("A:A", 12)
    ws_candidates.set_column("B:B", 22)
    ws_candidates.set_column("C:C", 88)
    ws_candidates.set_column("D:G", 13)
    ws_candidates.freeze_panes(4, 0)

    # Sheet 2: formula-driven EO validation.
    ws_validation = workbook.add_worksheet("EO Validation")
    ws_validation.hide_gridlines(2)
    ws_validation.merge_range(
        "A1:E1", "Tahap 2. Penerapan Persamaan Entropy Overlap (EO)", title_fmt
    )
    ws_validation.merge_range(
        "A2:E2",
        "EO dihitung dengan kontribusi per dokumen: ((-1/f_d) * LN(1/f_d)); f_d adalah jumlah kandidat cluster yang memuat dokumen D tersebut.",
        subtitle_fmt,
    )
    ws_validation.write("A4", "Frekuensi dokumen f_d")
    ws_validation.write("A5", "Kontribusi EO per dokumen")

    doc_headers = [f"D{doc_id}" for doc_id in range(1, doc_count + 1)]
    for col_offset, doc_header in enumerate(doc_headers):
        col = doc_start_col + col_offset
        col_name = excel_col_name(col)
        doc_id = col_offset + 1
        ws_validation.write(3, col, doc_header, header_fmt)
        ws_validation.write_formula(
            3,
            col,
            f"=SUM({col_name}${candidate_start_excel_row}:{col_name}${candidate_end_excel_row})",
            integer_fmt,
            doc_frequencies[doc_id],
        )
        ws_validation.write_formula(
            4,
            col,
            f"=IF({col_name}$4>1,((-1/{col_name}$4)*LN(1/{col_name}$4)),0)",
            formula_fmt,
            doc_contributions[doc_id],
        )

    ws_validation.write_row(
        6,
        0,
        ["No Cluster", "Frequent term set", "Cluster candidate", "Support", "EO Excel", *doc_headers],
        header_fmt,
    )

    for idx, (row, member_docs) in enumerate(zip(iteration_one, memberships), start=0):
        excel_row = candidate_start_excel_row + idx
        row_idx = excel_row - 1
        ws_validation.write_number(row_idx, 0, as_int(row["no_cluster"]), integer_fmt)
        ws_validation.write(row_idx, 1, row["frequent_term_set"], text_fmt)
        ws_validation.write(row_idx, 2, row["cluster_candidate"], text_fmt)
        ws_validation.write_formula(
            row_idx,
            3,
            f"=SUM({doc_start_name}{excel_row}:{doc_end_name}{excel_row})",
            formula_fmt,
            len(member_docs),
        )
        ws_validation.write_formula(
            row_idx,
            4,
            f"=SUMPRODUCT({doc_start_name}{excel_row}:{doc_end_name}{excel_row},${doc_start_name}$5:${doc_end_name}$5)",
            formula_fmt,
            excel_eo_values[idx],
        )
        for doc_id in range(1, doc_count + 1):
            ws_validation.write_number(
                row_idx,
                doc_start_col + doc_id - 1,
                1 if doc_id in member_docs else 0,
                member_fmt,
            )

    ws_validation.set_column("A:A", 12)
    ws_validation.set_column("B:B", 22)
    ws_validation.set_column("C:C", 88)
    ws_validation.set_column("D:E", 13)
    ws_validation.set_column(doc_start_col, doc_end_col, 5)
    ws_validation.freeze_panes(7, 5)

    # Sheet 3: selected clusters per FTC iteration.
    ws_selected = workbook.add_worksheet("Selected Clusters")
    ws_selected.hide_gridlines(2)
    ws_selected.merge_range("A1:F1", "Cluster Terpilih FTC per Iterasi", title_fmt)
    ws_selected.write_row(
        "A4",
        ["Iteration", "Cluster ID", "Frequent termset", "Support", "Entropy Overlap", "Remaining docs after"],
        header_fmt,
    )
    for idx, row in enumerate(selected_rows, start=5):
        ws_selected.write_number(idx - 1, 0, as_int(row["iteration"]), integer_fmt)
        ws_selected.write_number(idx - 1, 1, as_int(row["cluster_id"]), integer_fmt)
        ws_selected.write(idx - 1, 2, row["frequent_termset"], text_fmt)
        ws_selected.write_number(idx - 1, 3, as_int(row["support"]), integer_fmt)
        ws_selected.write_number(idx - 1, 4, as_float(row["entropy_overlap"]), number_fmt)
        ws_selected.write_number(idx - 1, 5, as_int(row["n_remaining_docs_after"]), integer_fmt)

    ws_selected.set_column("A:B", 12)
    ws_selected.set_column("C:C", 28)
    ws_selected.set_column("D:F", 18)
    ws_selected.freeze_panes(4, 0)

    workbook.close()
    return WORKBOOK_PATH


def verify_workbook(path: Path) -> None:
    workbook = openpyxl.load_workbook(path, data_only=False, read_only=True)
    expected_sheets = {"FTC Candidates", "EO Validation", "Selected Clusters"}
    missing_sheets = expected_sheets - set(workbook.sheetnames)
    if missing_sheets:
        raise RuntimeError(f"Sheet tidak ditemukan: {sorted(missing_sheets)}")

    validation = workbook["EO Validation"]
    candidates = workbook["FTC Candidates"]

    if validation["E8"].data_type != "f":
        raise RuntimeError("Formula EO Validation!E8 tidak ditemukan.")
    if candidates["F5"].data_type != "f":
        raise RuntimeError("Formula FTC Candidates!F5 tidak ditemukan.")

    workbook.close()


if __name__ == "__main__":
    result_path = build_workbook()
    verify_workbook(result_path)
    print(result_path)
